internal/compiler/pathutils.rs

```diff
@@ -0,0 +1,528 @@
+                PC::EmptyComponent | PC::SameDirectoryComponent(_) => {
+                    unreachable!("Never in the vector!")
+                }
+                PC::ParentDirectoryComponent(v) => {
+                    result += &format!("{v}{separator}");
+                }
+                PC::DirectoryComponent(v) => result += &format!("{v}{separator}"),
+                PC::FileComponent(v) => {
+                    result += v;
+                }
+            }
+        }
+        result
+    }
+}
+
+pub fn dirname_string(path: &str) -> String {
+    let mut iter = component_iter(path);
+    let separator = iter.separator();
+    let mut result = String::new();
+
+    while let Some(component) = iter.next() {
+        match component {
+            PathComponent::RootComponent(v) => result = v.to_string(),
+            PathComponent::EmptyComponent => result.push(separator),
+            PathComponent::SameDirectoryComponent(v)
+            | PathComponent::ParentDirectoryComponent(v)
+            | PathComponent::DirectoryComponent(v) => result += &format!("{v}{separator}"),
+            PathComponent::FileComponent(_) => { /* nothing to do */ }
+        };
+    }
+
+    if result.is_empty() {
+        String::from(".")
+    } else {
+        result
+    }
+}
+
+pub fn dirname(path: &Path) -> PathBuf {
+    let Some(path_str) = path.to_str() else {
+        // URLs can always convert to string in Rust
+        return path.to_owned();
+    };
+
+    PathBuf::from(dirname_string(path_str))
+}
+
+pub fn clean_path(path: &Path) -> PathBuf {
+    let Some(path_str) = path.to_str() else {
+        // URLs can always convert to string in Rust
+        return path.to_owned();
+    };
+
+    if is_url(path) {
+        // URL is cleaned up while parsing!
+        let url = url::Url::parse(path_str).unwrap();
+        return PathBuf::from(url.to_string());
+    }
+
+    PathBuf::from(clean_path_string(path_str))
+}
+
+pub fn join(base: &Path, path: &Path) -> Option<PathBuf> {
+    if is_absolute(path) {
+        return Some(path.to_owned());
+    }
+
+    let Some(base_str) = base.to_str() else {
+        // URLs can always convert to string in Rust
+        return Some(path.to_owned());
+    };
+    let Some(path_str) = path.to_str() else {
+        // URLs can always convert to string in Rust
+        return Some(path.to_owned());
+    };
+
+    if is_url(base) {
+        let base_url = url::Url::parse(base_str).unwrap();
+        Some(PathBuf::from(base_url.join(path_str).ok()?.to_string()))
+    } else {
+        let Some((_, _, separator)) = components(base_str, 0, &None) else {
+            return None;
+        };
+        let joined =
+            clean_path_string(&format!("{}{separator}{path_str}", dirname_string(base_str)));
+        Some(PathBuf::from(joined))
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    use std::path::PathBuf;
+
+    #[test]
+    fn test_is_url() {
+        fn th(input: &str, expected: bool) {
```

Comment:
you can use `#[track_caller]` for such funciton so the test will tell you which line actually failled.